feat(core): support injection token as predicate in queries

devversion opened the pull request:
Currently Angular internally already handles `InjectionToken` as
predicates for queries. This commit exposes this as public API as
developers already rely on this functionality but currently use
workarounds to satisfy the type constraints (e.g. `as any`).

We intend to make this public as it's low-effort to support, and
it's a significant key part for the use of light-weight tokens as
described in the upcoming guide: #36144.

In concrete, applications might use injection tokens over classes
for both optional DI and queries, because otherwise such references
cause classes to be always retained. This was also an issue in View
Engine, but now with Ivy, this pattern became worse, as factories are
directly attached to retained classes (ultimately ending up in the
production bundle, while being unused).

More details in the light-weight token guide and in: angular/angular-cli#16866.

Closes #21152. Related to #36144.

jelbourn reviewed (approved):
LGTM

IgorMinar commented on goldens/public-api/core/core.d.ts:
```diff
@@ -163,11 +163,11 @@ export declare interface ConstructorSansProvider {
 
 export declare interface ContentChildDecorator {
-    (selector: Type<any> | Function | string, opts?: {
+    (selector: Type<any> | InjectionToken<any> | Function | string, opts?: {
```
Any reason not to use `unknown` instead of `any`? Historically we used to use `any` in our public api surface whenever the type was unknown, but ever since TypeScript introduced the `unknown` type we started to to use that one instead. Sometimes `unknown` is not suitable, but in this case it might be. Can you please try it?


```suggestion
    (selector: Type<any> | InjectionToken<unknown> | Function | string, opts?: {
```

devversion replied:
good idea! no strong reason. I just did it that way to align with existing generics for query decorators.  Updated it to use `unknown`!

I assume we can clean up the other generics too. I can try sending a follow-up, or incorporate this here, if possible.

IgorMinar reviewed (changes requested):
otherwise LGTM

Reviewed-for: public-api

devversion committed:
feat(core): support injection token as predicate in queries

Currently Angular internally already handles `InjectionToken` as
predicates for queries. This commit exposes this as public API as
developers already relied on this functionality but currently use
workarounds to satisfy the type constraints (e.g. `as any`).

We intend to make this public as it's low-effort to support, and
it's a significant key part for the use of light-weight tokens as
described in the upcoming guide: https://github.com/angular/angular/pull/36144.

In concrete, applications might use injection tokens over classes
for both optional DI and queries, because otherwise such references
cause classes to be always retained. This was also an issue in View
Engine, but now with Ivy, this pattern became worse, as factories are
directly attached to retained classes (ultimately ending up in the
production bundle, while being unused).

More details in the light-weight token guide and in: https://github.com/angular/angular-cli/issues/16866.

Closes #21152. Related to #36144.

IgorMinar reviewed (approved):
LGTM now! Thanks

(changing existing apis to use `unknown` could be a breaking change, so we can do so only in a major version and only with proper discussion/review. So I suggest we punt that discussion until master is open for v11)

Reviewed-for: global-approvers

[presubmit](http://test/OCL:315949641:BASE:315949698:1591902482391:17d3956c)
